### PR TITLE
Updating sir and pkpd model to new ODE interface

### PR DESCRIPTION
**Pull request by bbbales2**
My original intention was just to convert the sir and pkpd models to the new interface so that the links in [here](https://github.com/stan-dev/example-models/pull/183) could be updated. sir was easy to convert, but the model in the pkpd folder didn't have the tests set up (to check if I made the changes correctly -- which is in the spirit of this repo) and so what I did was:

1. Generate empirical results ([here](https://github.com/stan-dev/stat_comp_benchmarks/blob/1027a5aa33c105bd26a1402a5ac0b2dcb0255ea7/empirical_results/nuts/one_comp_mm_elim_abs.fit)) using the old model with the sampler. I didn't generate vb results cause I didn't want to take the time to make sure the vb was working correctly, but the sampler results are there.

2. Rename the pkpd folder to one_comp_mm_elim_abs (so the test script would work well)

3. Removed the sim_one_comp_mm_elim_abs model (which is just simulating new data and was violating the one script per folder thing)

4. Updated sir and one_comp_mm_elim_abs models to use the new variadic ode interface (design-doc #19)

5. Remove the old run script and replace it with something much simpler (like what sir has)

### CODE REVIEW

**Comment by wds15** [comment]
Should we move to the "tol" solvers and explicitly specify tolerances and all that? I would find that much better and it should be more robust in case we adjust these in the future.

**Comment by bbbales2** [comment]
Sure, I'm fine with that.

**Comment by bbbales2** [comment]
@wds15 _tol solvers in place

**Comment by bbbales2** [comment]
@wds15 this needs merged before I can finish up the case study (so the links in the case study are stable).

**Comment by wds15** [comment]
Will finish this tomorrow...but I thought everything was fine, but I can confirm with a fresh mind.

File: benchmarks/one_comp_mm_elim_abs/one_comp_mm_elim_abs.info [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Reference calculations made with
+
+./one_comp_mm_elim_abs sample num_samples=1000 adapt delta=0.9 data file=one_comp_mm_elim_abs.data.R id=1 output file=output.1.csv
+./one_comp_mm_elim_abs sample num_samples=1000 adapt delta=0.9 data file=one_comp_mm_elim_abs.data.R id=2 output file=output.2.csv
```
Comment (wds15) [inline benchmarks/one_comp_mm_elim_abs/one_comp_mm_elim_abs.info]: How about fixing the random seed?

File: benchmarks/one_comp_mm_elim_abs/one_comp_mm_elim_abs.stan [hunk 2]
```diff
@@ -49,11 +42,9 @@ parameters {
...
 
 transformed parameters {
-  real C[N_t, 1];
-  {
-    real theta[3] = {k_a, K_m, V_m};
-    C = integrate_ode_bdf(one_comp_mm_elim_abs, C0, t0, times, theta, x_r, x_i);
-  }
+  vector[1] C[N_t] = ode_bdf_tol(one_comp_mm_elim_abs, C0, t0, times,
+				 1e-8, 1e-8, 1000,
+				 k_a, K_m, V_m, D, V);
```
Comment (wds15) [inline benchmarks/one_comp_mm_elim_abs/one_comp_mm_elim_abs.stan:47]: Can you please vectorise the likelihood? It wasn't vectorised before, but it should have been. Also: can we please swap `C_hat` to be the mean while `C` is the observed thing?  So maybe write

`C ~ lognormal(log(C_hat[,1]), sigma); `

and swap the declarations for data / transformed data. Calling C_hat the observed (with noise) is really odd to me. Thanks.

File: benchmarks/sir/sir.stan [hunk 3]
```diff
@@ -66,6 +56,5 @@ model {
   for (n in 2:N_t)
     stoi_hat[n] ~ poisson(y[n - 1, 1] - y[n, 1]);
 
-  B_hat ~ lognormal(log(col(to_matrix(y), 4)), 0.15);
-  
+  B_hat ~ lognormal(log(y[, 4]), 0.15);
```
Comment (wds15) [inline benchmarks/sir/sir.stan]: Can we also vectorise the for-loop for stoi_hat please? And get the naming right.. so "hat" is the mean always while the observed data is not "hat"?

The for loop probably needs storing for a temporary for the differenced y.

Thanks.

**Review by wds15 — CHANGES_REQUESTED** [review]
Sorry, can we please correct the convention thing with "hat" and also use properly vectorisation here? This way we really focus the performance test on the ODE stuff and these codes are closer to a reference (as fast as possible) implementation.

File: benchmarks/one_comp_mm_elim_abs/one_comp_mm_elim_abs.info (hunk 1 above)
Comment (bbbales2) [reply benchmarks/one_comp_mm_elim_abs/one_comp_mm_elim_abs.info]: Seeds are misleading for this stuff. It should work regardless of the seeds.

File: benchmarks/sir/sir.stan (hunk 3 above)
Comment (bbbales2) [reply benchmarks/sir/sir.stan]: This would be the vectorized code:

```
stoi[1] ~ poisson(y0[1] - y[1, 1]);
stoi[2:N_t] ~ poisson(y[1:(N_t - 1), 1] - y[2:N_t, 1]);
```

It doesn't work because `y[1:(N_t - 1), 1]` is an array. We could store that in a temporary, but I like the for loop better so I'm staying with a for loop.

Comment (bbbales2) [reply benchmarks/sir/sir.stan]: Aww can we keep the hats? If I take out the hats then the data is different from the new and the old model. That's just inconvenient.

Comment (bbbales2) [reply benchmarks/sir/sir.stan]: Ah it's easier to just do this than argue about it. I'll change the data. It adds an extra link to the new interface case study but I suppose we'll survive :P.

File: benchmarks/one_comp_mm_elim_abs/one_comp_mm_elim_abs.info (hunk 1 above)
Comment (wds15) [reply benchmarks/one_comp_mm_elim_abs/one_comp_mm_elim_abs.info]: but different seeds can really have a big impact on speed things and the sampler results unless you sample a crazy amount of iterations. If you don't want seeds, then at least lets fix `init=0` (if you think that's not good, then a small number is also fine).

... also, I think we fix seeds for all the other examples, so we should be consistent would be my position here. I would prefer the seeds to be set here. Maybe we run this by discourse quickly?

The thing is that seeds will get the variation from run to run down - and that's a good thing for this type of test which we want to run as reproducible as possible.

File: benchmarks/sir/sir.stan (hunk 3 above)
Comment (wds15) [reply benchmarks/sir/sir.stan]: Thanks for getting the hats replaced... the for loop is still the far worse way of writing this. What's so bad about

```stan
  {
   vector[N_t] y_diff;
    y_diff[1] = y0[1] - y[1, 1];
    for (n in 2:N_t)
        y_diff[n] = y[n - 1, 1] - y[n, 1];

   stoi ~ poisson(y_diff);
  }
```

This is a lot faster for sure.

... but... OK, I just saw that N_t is 20... so whatever you prefer (I just think that these codes should reflect good Stan coding practice).

**Comment by betanalpha** [comment]
My reply emails are unfortunately getting bounced so I'm posting directly from the web interface.

I’m confused about the status of this pull request and who has privileges to accept it.

More importantly this repository is meant to test the _statistical accuracy_ of an 
algorithm on a very mild suite of target distributions and has nothing to do with 
performance.  Consequently no great effort has been made to optimize the Stan 
programs and none of the Stan programs need to change unless there are breaking 
changes to the Stan language; given that the old format for the ODE functions is
remaining backward compatible there is zero need to change anything.

The programs in this repository were co-opted for the “performance tests” in 
continuous interaction, which are really regression tests, but they are defined 
independently of those tests.  If the performance tests need to change then a 
new repository should be set up instead of changing this repository for something 
it was never intended for.  Although in this case speed changes, either positive
or negative, don’t actually affect the performance tests as the gold values will 
have to be updated anyways.

Because it has come up in the comments — in the statistical literature the \hat
decoration is reserved for proper statistics, i.e. functions of the data.  They are
not used to denote functions of the parameters.  The confusing use of terms 
like mu_hat in some applied fields to denote location parameters in a model 
arose out of historical confusion/sloppiness/etc and should not be adopted here.

Overall I see no need to change anything in this repository.  I don’t think the 
continuous integration models need to be changed either, but I have less strong
opinions about that.

**Comment by wds15** [comment]
As a matter of fact these models are being run a lot during our CI for the purpose of performance testing. It sounds as if this is an abuse of these models for this purpose from you perspective... ok... but in any case we are going to save a lot of computational resources in going forward as these models are run over and over. We can run them ~30% faster and get the same result. So less energy wasted, more Greta points for us.

The hat notation for raw data (as it looked to me) is to me very confusing.

**Comment by bbbales2** [comment]
> who has privileges to accept it.

It's a stan-dev repo so it's part of the Stan project. @wds15 was willing to review it and I'm pretty sure is qualified enough.

> none of the Stan programs need to change unless there are breaking changes to the Stan language; given that the old format for the ODE functions is remaining backward compatible there is zero need to change anything.

The old interfaces are deprecated. We'll be getting rid of them and everything will convert eventually. These were convenient to convert because they're used in the performance tests and @rok-cesnovar suggested we make the changes. I wanted to wait on rstan to get the new interfaces before converting the case studies.

> Although in this case speed changes, either positive or negative, don’t actually affect the performance tests as the gold values will have to be updated anyways.

This repository doesn't have golds? Or are you referencing the empirical_results?

Anyway, this pull request also updates the model that was in the pkpd folder to actually be part of the tests here (it isn't currently even tested in develop).

> The confusing use of terms like mu_hat in some applied fields to denote location parameters in a model

The hats were being used like data. Maybe these are mean estimates passed in as data, I'm not sure. I glanced through the paper attached to the SIR model and didn't see any hats. I don't think there was a reference for the other model.

File: benchmarks/sir/sir.stan (hunk 3 above)
Comment (bbbales2) [reply benchmarks/sir/sir.stan]: Yeah you're right this would be better. Fixed.

File: benchmarks/one_comp_mm_elim_abs/one_comp_mm_elim_abs.info (hunk 1 above)
Comment (bbbales2) [reply benchmarks/one_comp_mm_elim_abs/one_comp_mm_elim_abs.info]: I put seeds back in there for these two models.

I suspect the performance tests are using different run scripts than were here (sir here was using regular non-NUTS hmc). I think we can probably get seeds in whatever scripts are being used there.

**Review by wds15 — APPROVED** [review]

**Comment by andrewgelman** [comment]
Hi, just to clarify on something that Michael wrote in his second-to-last-paragraph below:  In the statistical literature, the "hat" decoration is used for functions of data, but it is also used more generally for functions of parameters and data.  There are many examples; for one, see the Bugs code on page 360 of my book with Jennifer Hill on multilevel modeling.

There is no unique notation for regression models in the statistical literature.  Indeed, in our book we have a section, "Five ways to write the same model."

For further background on the general idea, in our 1996 paper on posterior predictive checks, Meng, Stern, and discussed the generalization of "test statistics" (functions of the data alone) to "realized discrepancies" (functions of data and parameters).   The idea of generalizing from functions of data to functions of data and parameters is not "historical confusion/sloppiness/etc."  It has a history in statistics--indeed, the idea is very Bayesian.

In short, it's ok in Stan models to use terms like mu_hat to denote location parameters.  More generally, we need to think harder on establishing clear notation and conventions for hierarchical models.  I think this is a hard problem that we haven't figured out yet.

Andrew

**Comment by bob-carpenter** [comment]
> Overall I see no need to change anything in this repository.

Sure, this could be duplicated, but what's going to be easiest for the project to maintain?  

Given that we just deprecated the current model formats, that seems like a strong motivation to update these.

> I’m confused about the status of this pull request and who has privileges to accept it.

Not surprising given that we currently have no technical governance for dealing with PRs for which there is not consensus.   This is why your own PR is stuck on the MCMC stuff.  

Soon, we should have a voting scheme to deal with this and the answer will be anyone with voting privilege will be able to vote at their own discretion.

What we're moving away from is having "owners" of repositories who have privileged positions w.r.t. managing PRs.   For devs who want that level of control, they can make repos in their own organizations.

>  The confusing use of terms like mu_hat in some applied fields to denote location parameters in a model arose out of historical confusion/sloppiness/etc and should not be adopted here.

Completely agree here.  We see this kind of thing a lot where they're used as local variables, not parameters:

```
for (n in 1:N) {
  real mu_hat = alpha + beta * x[n];
  y[n] ~ normal(mu_hat, sigma);
}
```

I really dislike overloading notation this way because `mu_hat` is not an estimate here.  So this just feels notationally wrong to me and thus I avoid it all cost so as not to confuse readers.  

@andrewgelman writes:
> In the statistical literature, the "hat" decoration is used for functions of data, but it is also used more generally for functions of parameters and data.  There are many examples; for one, see the Bugs code on page 360 of my book with Jennifer Hill on multilevel modeling.

I wouldn't think "in the statistical literature" would be much of an argument for a notational iconoclast :-). I also don't think self-citations count when arguing something is standard practice in the stats literature!

> ... In short, it's ok in Stan models to use terms like mu_hat to denote location parameters.

We see authors writing `p-hat(...)` in machine learning, but we presumably don't want to follow that precedent either.   If you look in our example-models repo, the only place the `_hat` convention is used is in translation of the ARM models from the first edition of your book (and one I did translating from the ML literature).

Wikipedia defines the "hat operator" as an estimator, as does their page on estimators, as does Wolfram and every every other reference or explanation I can find of what "hat" means in statistics.

> My reply emails are unfortunately getting bounced so I'm posting directly from the web interface.

My email replies to GitHub also bounce from gmail.   Does anyone know how to fix that?
